include/matx/transforms/solve/solve_cusparse.h

```diff
@@ -0,0 +1,186 @@
+  }
+  MATX_ASSERT(ret == CUSPARSE_STATUS_SUCCESS, matxSolverError);
+
+  matxFree(workspace);
+
+  ret = cusparseDestroy(handle);
+  MATX_ASSERT(ret == CUSPARSE_STATUS_SUCCESS, matxSolverError);
+}
+
+template <typename Op>
+__MATX_INLINE__ auto getCuSparseSolveSupportedTensor(const Op &in,
+                                                     cudaStream_t stream) {
+  const auto func = [&]() {
+    if constexpr (is_tensor_view_v<Op>) {
+      return in.Stride(Op::Rank() - 1) == 1;
+    } else {
+      return true;
+    }
+  };
+  return GetSupportedTensor(in, func, MATX_ASYNC_DEVICE_MEMORY, stream);
+}
+
+} // end namespace detail
+
+template <typename TensorTypeC, typename TensorTypeA, typename TensorTypeB>
+void sparse_dia_solve_impl(TensorTypeC &C, const TensorTypeA &a,
+                           const TensorTypeB &B, const cudaExecutor &exec) {
+  MATX_NVTX_START("", matx::MATX_NVTX_LOG_API)
+  const auto stream = exec.getStream();
+
+  // Transform into supported form.
+  auto b = getCuSparseSolveSupportedTensor(B, stream);
+  auto c = getCuSparseSolveSupportedTensor(C, stream);
+  if (!is_matx_transform_op<TensorTypeB>() && !b.isSameView(B)) {
+    (b = B).run(stream);
+  }
+
+  using atype = TensorTypeA;
+  using btype = decltype(b);
+  using ctype = decltype(c);
+
+  using TA = typename atype::value_type;
+  using TB = typename btype::value_type;
+  using TC = typename ctype::value_type;
+
+  static constexpr int RANKA = atype::Rank();
+  static constexpr int RANKB = btype::Rank();
+  static constexpr int RANKC = ctype::Rank();
+
+  // Restrictions.
+  static_assert(RANKA == 2 && RANKB == 2 && RANKC == 2,
+                "tensors must have rank-2");
+  static_assert(std::is_same_v<TC, TA> && std::is_same_v<TC, TB>,
+                "tensors must have the same data type");
+  static_assert(std::is_same_v<TC, float> || std::is_same_v<TC, double> ||
+                    std::is_same_v<TC, cuda::std::complex<float>> ||
+                    std::is_same_v<TC, cuda::std::complex<double>>,
+                "unsupported data type");
+  MATX_ASSERT(                                  // Note: B,C transposed!
+      a.Size(RANKA - 1) == a.Size(RANKA - 2) && // square
+          a.Size(RANKA - 1) == b.Size(RANKB - 1) &&
+          a.Size(RANKA - 2) == c.Size(RANKC - 1) &&
+          b.Size(RANKB - 2) == c.Size(RANKC - 2),
+      matxInvalidSize);
+  MATX_ASSERT(b.Stride(RANKB - 1) == 1 && c.Stride(RANKC - 1) == 1,
+              matxInvalidParameter);
+
+  if constexpr (atype::Format::isDIAI()) {
+    // These are *run-time* checks.
+    if (!c.isSameView(b)) {
+      MATX_THROW(matxNotSupported, "Tridiagonal solve overwrites rhs");
+    }
+    using CRD = typename atype::crd_type;
+    CRD *diags = a.CRDData(1);
+    const index_t numD = a.crdSize(1);
+    if (numD != 3 || diags[0] != -1 || diags[1] != 0 || diags[2] != 1) {
+      MATX_THROW(matxNotSupported, "Only tridiagonal solve supported");
+    }
+    using T = std::conditional_t<
```

Comment:
This would be a bit easier to read with a type trait, but we can leave it for now.